Add support for other authentication methods in AWS S3

stujonesuk opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This change allows users of the AWS S3 integration to select from 3 authentication methods:

* Implicit Authentication (e.g. IAM Instance Profile, IAM Roles Anywhere, default AWS Config profile auth)
* Named Profile (non-default AWS Config profile auth)
* IAM Credentials (explicit IAM User Access Key ID and Secret Key) - currently the only supported authentication type.

The addition of the two new authentication types allows the AWS credentials management to be delegated outside of the control of this specific integration.

Additionally, the reauth and reconfigure config flows have been added, and the diagnostics flow has been added.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: https://github.com/orgs/home-assistant/discussions/1281
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/41281
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [X] I understand the code I am submitting and can explain how it works.
- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.
- [X] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

stujonesuk commented:
> This is very similar to #146475, which was left stale after [my comment](https://github.com/home-assistant/core/issues/145862#issuecomment-2960486992). What do you think about implementing it as described there?

Apologies, this existing PR didn't come up when I searched for similar existing issues, obviously didn't get the search terms right!

I've not looked at how to do multi-step config, but I'll give it a go tonight.  I'll drop this back to draft status for now.

tomasbedrich commented:
Thank you! _But_ - there is [another opinion](https://github.com/home-assistant/core/issues/145862#issuecomment-3378525367), which I have to agree with 🫠 For the sake of simplicity - what if we stick without your original concept? Just without "NONE" as a special value, but rather keeping the input empty = trying to call boto3 without params? Note that it all relies on a nice explanation in config flow description.

kennedyshead reviewed (changes requested):
This PR is way to big to review! Could you split it up somehow?

stujonesuk committed:
Add new Authentication Mode and Profile Name constants

stujonesuk committed:
Add Configuration Model base configuration

stujonesuk committed:
Add Configuration Model dictionary helper methods

Including ability to filter output dictionary to a specified set of keys

stujonesuk committed:
Add Configuration Model Error Management Methods

stujonesuk committed:
Add Configuration Model Authentication Mode Logic

stujonesuk committed:
Add Configuration Model Validation Methods

stujonesuk committed:
Switch Integration Init to use new Config Model

stujonesuk committed:
Add Named Profile support to Integration Initialization

stujonesuk committed:
Add support for additional error types and reauth flow to Integration Initialisation

stujonesuk committed:
Ensure boto3 client id disposed in the event of errors in the Integration Initialisation

stujonesuk committed:
Add strings to split config flow into three (bucket, profile, explicit)

stujonesuk committed:
Major changes to Config Flow

1. Switch to using new Config Model (S3ConfigModel) under the hood that persists for lifetime of the flow
2. Split into three strps - bucket, profile, explicit
3. Support editing flows - including read-only fields,
4. Support bouncing back to bucket step if errors that can only be resolved at that step are raised

stujonesuk committed:
Add reauth and reconfigure triggers to config flow

stujonesuk committed:
Tests - clear down orphaned config models at the end of each test

stujonesuk committed:
Tests - define test inputs for different use cases

stujonesuk committed:
Tests - refactor Init tests to cover additional error types, and scenarios

Cover errors raised at session creation, client creation, and bucket head

stujonesuk committed:
Tests - remove deprecated tests from old config flow

stujonesuk committed:
Tests - add helper functions

stujonesuk committed:
Tests - validate initial state for config flow triggered by user, reauth, or reconfigure

stujonesuk committed:
Tests - confirm flow termination cleans up orphaned models

stujonesuk committed:
Tests - user-triggered flow, implicit auth, success

stujonesuk committed:
Tests - user-triggered flow, profile auth, success

stujonesuk committed:
Tests - user-triggered flow, explicit auth, success

stujonesuk committed:
Tests - user-triggered flow, errors

stujonesuk committed:
Tests - reauth-triggered flow, implicit auth, success

stujonesuk committed:
Tests - reauth-triggered flow, profile auth, success

stujonesuk committed:
Tests - reauth-triggered flow, explicit auth, success

stujonesuk committed:
Tests - reauth-triggered flow, errors

stujonesuk committed:
Tests - reconfigure-triggered flow, implicit auth, success

stujonesuk committed:
Tests - reconfigure-triggered flow, profile auth

stujonesuk committed:
Tests - reconfigure-triggered flow, explicit auth

stujonesuk committed:
Tests - reconfigure-triggered flow, errors

stujonesuk committed:
Tests - config model init

stujonesuk committed:
Tests - config model dictionary functions

stujonesuk committed:
Tests - config model, validate access, success

stujonesuk committed:
Tests - config model, validate access, various failures

stujonesuk committed:
Tests - config model, filter errors

stujonesuk committed:
Add Diagnostics flow support

stujonesuk committed:
Tests - Diagnostics

stujonesuk committed:
Quality Scale - update with new implemented flows

joostlek reviewed (changes requested):
I would argue that diagnostics, reauth, reconfigure and you change all should be their own PRs
